explainer.md

```diff
@@ -0,0 +1,273 @@
+  await stream.waitForReadable(4);
+  if (!stream.readable) {
+    return null;
+  }
+  const buffer = new Uint8Array(4);
+  const read = stream.readInto(buffer);
+  return new DataView(buffer.array).getUint32(0);
+}
+
+async function readStreamBytes(stream, count) {
+  const buffer = new ArrayBuffer(count);
+  let bufferedAmount = 0;
+  while (bufferedAmount < count) {
+    await stream.waitForReadable(1);
+    if (!stream.readable) {
+      return null;
+    }
+    const read = stream.readInto(new Uint8Array(buffer, bufferedAmount));
+    bufferedAmount += read.amount;
+  }
+  return buffer;
+}
+```
+
+## Example of receiving notifications pushed from the server, with responses
+
+```javascript
+const host = 'example.com';
+const port = 10001;
+const transport = new QuicTransport(host, port);
+
+// Note that the notifications will arrive out of order
+transport.onbidirectionalstream = (e) => {
+  const notification = readStreamUntilFin(e.stream)
+
+  if (notification) {
+    // App-specific notification encoding
+    const notificationMessage = decodeNotification(notification);
+    let n = new Notification(notificationMessage);
+    n.onclick = (e) => {
+      // App-specific click message encoding
+      const clickMessage = encodeClickMessage();
+      e.stream.write(clickMessage);
+    };
+  }
+}
+
+async function readStreamUntilFin(stream) {
+  const buffers = [];
+  let bufferedSize = 0;
+  let finished = false;
+  while (stream.readable) {
+    await stream.waitForReadable(1);
+    const buffer = new Uint8Array(stream.readableAmount);
+    const read = stream.readInto(readBuffer);
+    bufferedSize += read.amount;
+    finished = read.finished;
+  }
+  if (!finished) {
+    return null;
+  }
+
+  let joinedBuffer = new Uint8Array(bufferedSize);
+  let joinedSize = 0;
+  for (let buffer of buffers) {
+    joinedBuffer.set(buffer, joinedSize);
+    joinedSize += buffer.byteLength;
+  }
+  return joinedBuffer.array;
+}
+```
+
+## Example of requesting over HTTP and receiving media pushed out-of-order and unreliably over the same network connection
+
+```javascript
+const transport = PooledHttpTransport.getDatagramTransport();
+if (transport) {
+  await fetch('http://example.com/babyshark');
+  const datagrams = await transport.receiveDatagrams();
+  for (let data of datagrams) {
+    if (data) {
+      // Process the data
+    }
+  };
+}
+```
+
+## Detailed design discussion
+
+Any WebTransport can provide any of the following capabilities (mixins):
+
+: Unidirectional streams
+:: Indefintely long streams of bytes in one direction with back pressure applied
+to the sender when either the receiver can't read quickly enough or when
+constrained by network capacity/congestions.  Useful for sending messages that
+do not expect a response.  In-order, reliable messaging can be achieved by
+sending many messages in a single stream. Out-of-order messaging can be achieved
+by sending one message per stream. unreliable messaging can be achieved by
```

Comment:
nit: s/unreliable/Unreliable/

Also maybe we should remove this sentence from the explainer since we're still not sure there's demand for it in the protocol where we have datagrams?

Reply:
OK, removed.